server/core_purchase.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2017 The Nakama Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"database/sql"
+	"nakama/pkg/iap"
+
+	"errors"
+
+	"encoding/json"
+
+	"github.com/satori/go.uuid"
+	"go.uber.org/zap"
+)
+
+type PurchaseService struct {
+	logger       *zap.Logger
+	db           *sql.DB
+	AppleClient  *iap.AppleClient
+	GoogleClient *iap.GoogleClient
+}
+
+func NewPurchaseService(jsonLogger *zap.Logger, multiLogger *zap.Logger, db *sql.DB, config *PurchaseConfig) *PurchaseService {
+	ac, err := iap.NewAppleClient(config.Apple.Password, config.Apple.Production, config.Apple.TimeoutMs)
+	if err != nil {
+		multiLogger.Warn("Skip initialising Apple in-app purchase provider", zap.Error(err))
+	} else {
+		if config.Apple.Production {
+			multiLogger.Info("Apple in-app purchase environment is set to Production priority.")
+		} else {
+			multiLogger.Info("Apple in-app purchase environment is set to Sandbox priority.")
+		}
+		multiLogger.Info("Successfully initiated Apple in-app purchase provider.")
+	}
+
+	gc, err := iap.NewGoogleClient(config.Google.PackageName, config.Google.ServiceKeyFilePath)
+	if err != nil {
+		multiLogger.Warn("Skip initialising Google in-app purchase provider", zap.Error(err))
+	}
+
+	return &PurchaseService{
+		logger:       jsonLogger,
+		db:           db,
+		AppleClient:  ac,
+		GoogleClient: gc,
+	}
+}
+
+func (p *PurchaseService) ValidateApplePurchase(userID uuid.UUID, purchase *iap.ApplePurchase) *iap.PurchaseVerifyResponse {
+	r, appleReceipt := p.AppleClient.Verify(purchase)
+	if !r.Success {
+		return r
+	}
+
+	//TODO: Improvement - Process more than one in-app receipts
+	inAppReceipt := appleReceipt.InApp[0]
+	p.checkUser(userID, r, 1, inAppReceipt.TransactionID)
+
+	if r.Success && !r.SeenBefore {
+		err := p.savePurchase(userID.Bytes(), 1, inAppReceipt.ProductID, inAppReceipt.TransactionID, purchase.ReceiptData, r.Data)
```

Comment:
`Success` has already been decided here, but we haven't updated the ledger yet. If the ledger update then fails because of a concurrent verification for the same purchase both will return success to the client, right?